google/cloud/bigtable/data/_cross_sync/_decorators.py

```diff
@@ -0,0 +1,441 @@
+        if isinstance(ast_node, ast.Dict):
+            return {
+                cls._convert_ast_to_py(k): cls._convert_ast_to_py(v)
+                for k, v in zip(ast_node.keys, ast_node.values)
+            }
+        raise ValueError(f"Unsupported type {type(ast_node)}")
+
+
+class ConvertClass(AstDecorator):
+    """
+    Class decorator for guiding generation of sync classes
+
+    Args:
+        sync_name: use a new name for the sync class
+        replace_symbols: a dict of symbols and replacements to use when generating sync class
+        docstring_format_vars: a dict of variables to replace in the docstring
+        rm_aio: if True, automatically strip all asyncio keywords from method. If false,
+            only keywords wrapped in CrossSync.rm_aio() calls to be removed.
+        add_mapping_for_name: when given, will add a new attribute to CrossSync,
+            so the original class and its sync version can be accessed from CrossSync.<name>
+    """
+
+    def __init__(
+        self,
+        sync_name: str | None = None,
+        *,
+        replace_symbols: dict[str, str] | None = None,
+        docstring_format_vars: dict[str, tuple[str | None, str | None]] | None = None,
+        rm_aio: bool = False,
+        add_mapping_for_name: str | None = None,
+    ):
+        self.sync_name = sync_name
+        self.replace_symbols = replace_symbols
+        docstring_format_vars = docstring_format_vars or {}
+        self.async_docstring_format_vars = {
+            k: v[0] or "" for k, v in docstring_format_vars.items()
+        }
+        self.sync_docstring_format_vars = {
+            k: v[1] or "" for k, v in docstring_format_vars.items()
+        }
+        self.rm_aio = rm_aio
+        self.add_mapping_for_name = add_mapping_for_name
+
+    def async_decorator(self):
+        """
+        Use async decorator as a hook to update CrossSync mappings
+        """
+        from .cross_sync import CrossSync
+
+        if not self.add_mapping_for_name and not self.async_docstring_format_vars:
+            # return None if no changes needed
+            return None
+
+        new_mapping = self.add_mapping_for_name
+
+        def decorator(cls):
+            if new_mapping:
+                CrossSync.add_mapping(new_mapping, cls)
+            if self.async_docstring_format_vars:
+                cls.__doc__ = cls.__doc__.format(**self.async_docstring_format_vars)
+            return cls
+
+        return decorator
+
+    def sync_ast_transform(self, wrapped_node, transformers_globals):
+        """
+        Transform async class into sync copy
+        """
+        import ast
+        import copy
+
+        # copy wrapped node
+        wrapped_node = copy.deepcopy(wrapped_node)
+        # update name
+        if self.sync_name:
+            wrapped_node.name = self.sync_name
+        # strip CrossSync decorators
+        if hasattr(wrapped_node, "decorator_list"):
```

Comment:
is this looking for the string "decorator_list" in the class? Is there an example you can point me to in the other PR? I also don't see this in your doc https://docs.google.com/document/d/1cy5oq4UVyg96AF8tX9nZ0zw8pbyu_lgF1DfFFegrTk0/edit?resourcekey=0-glL8ejJV1gRNXseZwYVkvw&tab=t.0#heading=h.xzptrog8pyxf

Reply:
This is part of the `ast` stdlib. Some ast nodes contain a `.decorator_list` attribute we can access to see the decorators associated with the class or function: https://docs.python.org/3/library/ast.html#ast.FunctionDef